src/backend/egl/surface.rs

```diff
@@ -119,8 +115,11 @@ impl<N: native::NativeSurface> EGLSurface<N> {
                         SurfaceCreationError::NativeSurfaceCreationFailed(err) => {
                             SwapBuffersError::Underlying(err)
                         }
-                    })?
-            });
+                    })? as *mut _
+            }, Ordering::SeqCst);
+            if previous == surface && !surface.is_null() {
```

Comment:
Lets swap the order around to make sure, the old pointer cannot be accessed by anyone else anymore, before we destroy it. (`compare_and_swap` is perfect here and also makes sure, that parallel recreations don't override each other.)